Complete CRUD with ExerciseType entity

Kosiv92 opened the pull request.

ilyanikolaeff committed:
Исправление структуры проекта

Подключены дополнительные пакеты (Scrutor, Mediatr, FluentValidation)

ilyanikolaeff committed:
Added examples

ilyanikolaeff committed:
Added tests

Kosiv92 committed:
Fix route problems
Implement ExerciseTypeHandlers methods

ilyanikolaeff commented on src/IG_Train.Contracts/Responses/ExerciseTypeResponse.cs:
```diff
@@ -0,0 +1,10 @@
+﻿namespace IG_Train.Contracts.Responses;
+
+public record ExerciseTypeResponse
```
Сделай чере дефолт конструктор

ilyanikolaeff commented on src/IG_Train.Domain/Repositories/IRepository.cs:
```diff
@@ -0,0 +1,20 @@
+
+    public Task<TEntity?> GetByIdAsync(int id, CancellationToken cancellationToken = default);
+
+    public Task<TKey> CreateAsync(TEntity entity, CancellationToken cancellationToken = default);
```
Не делай default CancellationToken, пусть везде пробрасывается

ilyanikolaeff commented on src/IG_Train.Domain/Services/IExerciseTypeService.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    Task<int> CreateExerciseType(ExerciseTypeEntity exerciseType, CancellationToken cancellationToken = default);
+
+    Task DeleteExerciseType(int id, CancellationToken cancellationToken = default);
```
Аналогично с дефолт токеном

ilyanikolaeff commented on test/IG_Train.Tests.Unit/Application/Validators/CreateExerciseTypeRequestValidatorTests.cs:
```diff
@@ -0,0 +1,48 @@
+using IG_Train.Application.Validators;
+using Xunit;
+
+namespace IG_Train.Tests.Unit.Application.Validators
```
Тесты на свои сервисы

Kosiv92 committed:
Fix arguments in interfaces and their implementations. Move CQRS handlers to separate folders

Kosiv92 committed:
Delete src/IG_Train.Contracts directory

ilyanikolaeff commented on src/IG_Train.Application/Handlers/ExerciseType/Create/CreateExerciseTypeHandler.cs:
```diff
@@ -0,0 +1,21 @@
+    public async Task<CreateExerciseTypeResponse> Handle(CreateExerciseTypeRequest request, CancellationToken cancellationToken)
+    {
+        return new CreateExerciseTypeResponse(await _exerciseTypeService.CreateExerciseType
+            (new ExerciseTypeEntity(default, request.Name, request.Description), cancellationToken));        
```
Не стоит создание сущности пихать в хендлер, делегируй в сервис. Идеальный флоу хендлер -> сервис, при этом в хендлере ничего кроме перенаправления запроса

Kosiv92 replied:
В настоящий момент проект Domain не ссылается на проект Application, где определены request и response классы, соответственно тут либо добавлять из Domain ссылку на Application и тогда я из сервиса смогу возвращать необходимые объекты (что как я понимаю в корен неправильный подход), либо выделять еще одни сервис который будет заниматься созданием таких объектов

ilyanikolaeff commented on src/IG_Train.Application/Mapping/ApplicationMappingProfile.cs:
```diff
@@ -0,0 +1,13 @@
+{
+    public ApplicationMappingProfile()
+    {
+        CreateMap<UpdateExerciseTypeRequest, ExerciseTypeEntity>();
```
Хороший стиль маппинг расписывать по мемберам

ilyanikolaeff commented on src/IG_Train.Application/Validators/CreateExerciseTypeRequestValidator.cs:
```diff
@@ -0,0 +1,14 @@
+    {
+        RuleFor(x => x.Name)
+            .NotEmpty()
+            .WithMessage(x => $"{nameof(x.Name)} must not be null or empty");
```
Это я писал? 
Если я, то есть еще тема без интерполяции {PropertyName} в строке и FluentValidation автоматом проставит

ilyanikolaeff commented on src/IG_Train.Domain/Entities/ExerciseTypeEntity.cs:
```diff
@@ -0,0 +1,15 @@
+        Description = description;
+    }
+
+    public string Name { get; private set; }
```
Пока не ДДД можно без приватных сеттеров, но я не помню точно поддерживает ли EFCore такие приколы (вроде да)

Kosiv92 replied:
EF Core приватные сеттеры поддерживает. Я думаю, что если по какой-то причине не требуется обратное, то лучше ограничить запись новых значений в свойства

ilyanikolaeff commented on src/IG_Train.Domain/Services/ExerciseTypeService.cs:
```diff
@@ -0,0 +1,41 @@
+{
+    public class ExerciseTypeService : IExerciseTypeService
+    {
+        private readonly IRepository<ExerciseTypeEntity, int> _repositoryET;
```
Имя переменной шлак

ilyanikolaeff reviewed (changes requested):
Как пофиксишь отпиши общим комментом

Kosiv92 committed:
Add auto validation to pipeline
Implement unit tests for validators
Several fixes by pull request comments (dependency service name, mapping members, etc.)

Kosiv92 committed:
Delete comments block from tests
Delete unused dependency from tests

ilyanikolaeff commented on src/IG_Train.Application/Validators/CreateExerciseTypeRequestValidator.cs:
```diff
@@ -1,14 +1,17 @@
-            .NotEmpty()
-            .WithMessage(x => $"{nameof(x.Name)} must not be null or empty");
+        .NotNull()
+        .NotEmpty()
```
NotEmpty покрывает NotNull, можешь в доке убедиться

ilyanikolaeff commented on src/IG_Train.Web/Controllers/ExerciseTypeController.cs:
```diff
@@ -20,17 +23,17 @@ public ExerciseTypeController(IMediator mediator)
-            _mediator.Send(request, cancellationToken);
-
+            CancellationToken cancellationToken)
+            => await _mediator.Send(request, cancellationToken);
```
А в чем смысл async/await делать если можно просто прокинуть, без лишней инициализации стейт машины. Перформанс минимальный, а букв меньше

Kosiv92 replied:
В дальнейшем я могу добавить еще поля для этой сущности, добавить подгрузку картинки или еще как то изменить реализацию таким образом, что блокировка потока будет занимать заметное время. Понимаю, что возможно в данном случае я ничего не экономлю, но хотелось бы придерживаться единого стиля и по возможности использовать асинхронные операции

ilyanikolaeff replied:
Ну как будет, так и расширишь

ilyanikolaeff commented on src/IG_Train.Web/Extensions/ServiceCollectionExtensions.cs:
```diff
@@ -66,11 +68,12 @@ internal static IServiceCollection AddRepositories(this IServiceCollection servi
             typeof(IApplicationAssembly).Assembly,
             typeof(IInfrastructureAssembly).Assembly,
             typeof(IDomainAssembly).Assembly]);
+        services.AddFluentValidationAutoValidation();
```
Это точно из SharpGrip подключено? смущает юзинг на второй строке

ilyanikolaeff commented on test/IG_Train.Tests.Unit/Application/Validators/UpdateExerciseTypeRequestValidatorTests.cs:
```diff
@@ -0,0 +1,68 @@
+            var request = new UpdateExerciseTypeRequest(id, name, description);                
+
+            // Act
+            var result = sut.Validate(request);
```
TestValidateAsync

ilyanikolaeff commented on test/IG_Train.Tests.Unit/Application/Validators/UpdateExerciseTypeRequestValidatorTests.cs:
```diff
@@ -0,0 +1,68 @@
+            var request = new UpdateExerciseTypeRequest(default, string.Empty, string.Empty);
+
+            // Act
+            var result = sut.Validate(request);
```
Аналогично

Kosiv92 committed:
Fix unit tests. Delete unnecessary nuget. Fix rule in CreateExerciseTypeValidator

Kosiv92 committed:
Delete async/await from controllers methods
